### PR TITLE
[CDAP-15305] Introduce MutationOptions to control synchronous Metadata Storage

### PR DESCRIPTION
**Pull request by kmeelu**
JIRA: https://issues.cask.co/browse/CDAP-15305
BUILD: https://builds.cask.co/browse/CDAP-DRC9179-7

DBV: https://builds.cask.co/browse/CDAP-DBV35-1
ITM: https://builds.cask.co/browse/IT-ITM-189

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/MetadataHttpHandler.java [hunk 1]
```diff
@@ -142,7 +143,8 @@ public void getTags(HttpRequest request, HttpResponder responder,
   public void addProperties(FullHttpRequest request, HttpResponder responder,
                             @QueryParam("type") String type) throws BadRequestException, IOException {
     MetadataEntity metadataEntity = getMetadataEntityFromPath(request.uri(), type, "/metadata/properties");
-    metadataAdmin.addProperties(metadataEntity, readProperties(request));
+    MutationOptions options = MutationOptions.builder().setAsynchronous(false).build();
```
Comment (anew) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/MetadataHttpHandler.java]: seems that you could define these options as a 
```static final MutationOptions SYNC = MutationOptions.builder().setAsynchronous(false).build()```
and use that constant throughout this handler.

Comment (anew) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/MetadataHttpHandler.java]: even better, if we allow callers to specify ?async=true as part of the URL to override this.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/MetadataSubscriberServiceTest.java [hunk 2]
```diff
@@ -433,6 +434,7 @@ public void testProfileMetadata() throws Exception {
 
     // get the mds should be empty property since we haven't started the MetadataSubscriberService
     MetadataStorage mds = injector.getInstance(MetadataStorage.class);
+    MutationOptions options = MutationOptions.builder().build();
```
Comment (anew) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/MetadataSubscriberServiceTest.java]: use DEFAULT?

File: cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/MetadataSubscriberServiceTest.java [hunk 3]
```diff
@@ -580,6 +582,7 @@ public void testProfileMetadataWithNoProfilePreferences() throws Exception {
 
     // get the metadata - should be empty since we haven't deployed the app
     MetadataStorage mds = injector.getInstance(MetadataStorage.class);
+    MutationOptions options = MutationOptions.builder().build();
```
Comment (anew) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/MetadataSubscriberServiceTest.java]: ditto

File: cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/MetadataSubscriberServiceTest.java [hunk 4]
```diff
@@ -647,6 +650,7 @@ public void testAppDeletionMessage() throws Exception {
     // get the mds and put some workflow metadata in that, the publish of app deletion message should get the metadata
     // deleted
     MetadataStorage mds = injector.getInstance(MetadataStorage.class);
+    MutationOptions options = MutationOptions.builder().build();
```
Comment (anew) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/MetadataSubscriberServiceTest.java]: and same here: DEFAULT

File: cdap-data-fabric/src/test/java/io/cdap/cdap/data2/metadata/AuditMetadataStorageTest.java [hunk 5]
```diff
@@ -193,17 +194,24 @@ public void testPublishingDisabled() throws IOException {
...
+    storage.apply(new MetadataMutation.Update(dataset.toMetadataEntity(),
+                                              new io.cdap.cdap.spi.metadata.Metadata(MetadataScope.USER, datasetTags)),
+                  MutationOptions.DEFAULT);
     storage.apply(new MetadataMutation.Update(
-      dataset.toMetadataEntity(), new io.cdap.cdap.spi.metadata.Metadata(MetadataScope.USER, datasetTags)));
+      app.toMetadataEntity(), new io.cdap.cdap.spi.metadata.Metadata(MetadataScope.USER, appProperties)),
+                  MutationOptions.DEFAULT);
     storage.apply(new MetadataMutation.Update(
-      app.toMetadataEntity(), new io.cdap.cdap.spi.metadata.Metadata(MetadataScope.USER, appProperties)));
+      app.toMetadataEntity(), new io.cdap.cdap.spi.metadata.Metadata(MetadataScope.USER, appTags)),
```
Comment (anew) [inline cdap-data-fabric/src/test/java/io/cdap/cdap/data2/metadata/AuditMetadataStorageTest.java]: formatting is off here

File: cdap-metadata-spi/src/main/java/io/cdap/cdap/spi/metadata/MutationOptions.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+public class MutationOptions {
+
+  private final boolean asynchronous;
+
+  /**
+   * Create a mutation options object.
+   *
+   * @param asynchronous whether metadata mutation call should be asynchronous (non-blocking) or not.
+   */
+  public MutationOptions(boolean asynchronous) {
```
Comment (anew) [inline cdap-metadata-spi/src/main/java/io/cdap/cdap/spi/metadata/MutationOptions.java]: should be private to enforce use of the builder

File: cdap-metadata-spi/src/main/java/io/cdap/cdap/spi/metadata/MutationOptions.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  /**
+   * Builder for  mutation options.
+   */
+  public static class Builder {
+
+    private boolean asynchronous;
+
+    private Builder() {}
+
+    public Builder setAsynchronous(boolean asynchronous) {
```
Comment (anew) [inline cdap-metadata-spi/src/main/java/io/cdap/cdap/spi/metadata/MutationOptions.java:58]: you should document what the default is, if this is not called.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/MetadataSubscriberServiceTest.java (hunk 2 above)
Comment (kmeelu) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/MetadataSubscriberServiceTest.java]: Updating this file to use MutationOptions.DEFAULT.

**Review by anew — CHANGES_REQUESTED** [review]
Overall looks pretty good. One additional comment: I feel that we do not have tests that validate the new option.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/MetadataHttpHandler.java (hunk 1 above)
Comment (kmeelu) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/MetadataHttpHandler.java]: Yep, I've modified this and added the async QueryParam. PTAL.

File: cdap-data-fabric/src/test/java/io/cdap/cdap/data2/metadata/AuditMetadataStorageTest.java (hunk 5 above)
Comment (kmeelu) [reply cdap-data-fabric/src/test/java/io/cdap/cdap/data2/metadata/AuditMetadataStorageTest.java]: Tried fixing this. Let me know if there's a set standard we like to adhere to.

**Comment by kmeelu** [comment]
> Overall looks pretty good. One additional comment: I feel that we do not have tests that validate the new option.

I added the testAsyncMutations check. Batch test simply checks to make sure the changelists are the same, so it didn't seem to fit to have an async test; perhaps an integration test would be best to double check that logic.

File: cdap-metadata-spi/pom.xml [hunk 8]
```diff
@@ -36,6 +36,12 @@
       <artifactId>cdap-api</artifactId>
       <version>${project.version}</version>
     </dependency>
+    <dependency>
```
Comment (anew) [inline cdap-metadata-spi/pom.xml:39]: what is this needed for?

Comment (anew) [reply cdap-metadata-spi/pom.xml:39]: asking because we tried to keep the SPI dependencies to a minimum, ideally only depending on API.

Comment (anew) [reply cdap-metadata-spi/pom.xml:39]: oh it is only for tests, that is fine then.

**Review by anew — APPROVED** [review]

File: cdap-metadata-spi/pom.xml (hunk 8 above)
Comment (kmeelu) [reply cdap-metadata-spi/pom.xml:39]: Yep, added to enable Tasks.waitFor() in async tests within MetadataStorageTests.

**Pull request merged**